contracts/TwabRewards.sol

```diff
@@ -0,0 +1,392 @@
+     */
+    function _getCurrentEpochId(uint256 _promotionId) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+
+        // (_numberOfEpochs * elapsedTimestamp) / promotionDurationTimestamp
+        return
+            (_numberOfEpochs * (block.timestamp - _promotion.startTimestamp)) /
+            (_promotion.epochDuration * _numberOfEpochs);
+    }
+
+    /**
+        @notice Get reward amount for a specific user.
+        @dev Rewards can only be claimed once the epoch is over.
+        @param _user User to get reward amount for
+        @param _promotionId Promotion id from which the epoch is
+        @param _epochId Epoch id to get reward amount for
+        @return Reward amount
+     */
+    function _calculateRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        uint256 _epochDuration = _promotion.epochDuration;
+        uint256 _epochStartTimestamp = _promotion.startTimestamp + (_epochDuration * _epochId);
+        uint256 _epochEndTimestamp = _epochStartTimestamp + _epochDuration;
+
+        require(
+            block.timestamp >= _epochStartTimestamp && block.timestamp <= _epochEndTimestamp,
+            "TwabRewards/epoch-not-over"
+        );
+
+        ITicket _ticket = ITicket(_promotion.ticket);
+
+        uint256 _averageBalance = _ticket.getAverageBalanceBetween(
+            _user,
+            uint64(_epochStartTimestamp),
+            uint64(_epochEndTimestamp)
+        );
+
+        uint64[] memory _epochStartTimestamps = new uint64[](1);
+        _epochStartTimestamps[0] = uint64(_epochStartTimestamp);
+
+        uint64[] memory _epochEndTimestamps = new uint64[](1);
+        _epochEndTimestamps[0] = uint64(_epochEndTimestamp);
+
+        uint256[] memory _averageTotalSupplies = _ticket.getAverageTotalSuppliesBetween(
+            _epochStartTimestamps,
+            _epochEndTimestamps
+        );
+
+        return (_promotion.tokensPerEpoch * _averageBalance) / _averageTotalSupplies[0];
+    }
+
+    /**
+        @notice Get the total amount of tokens left to be rewarded.
+        @param _promotionId Promotion id to get the total amount of tokens left to be rewarded for
+        @return Amount of tokens left to be rewarded
+     */
+    function _getRemainingRewards(uint256 _promotionId) internal view returns (uint256) {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _numberOfEpochs = _promotion.numberOfEpochs;
+        uint256 _tokensPerEpoch = _promotion.tokensPerEpoch;
+        uint256 _currentEpochId = _getCurrentEpochId(_promotionId);
+
+        if (_currentEpochId == 0) {
+            return _tokensPerEpoch * _numberOfEpochs;
+        } else if (_numberOfEpochs <= _currentEpochId + 1) {
```

Comment:
This is confusing logic.

I think what you're trying to say is if the current epoch id is beyond the number of epochs, then it's zero.

This logic fails if, for example, _numberOfEpochs is 3 and _currentEpochId is 2.

Since number = 3, then we'll have epoch ids 0, 1, 2.  So epoch id 2 is valid.

3 <= 2 + 1 is true, however, so it'll return 0 for the last epoch.

Reply:
Yes, I've fixed it: https://github.com/pooltogether/v4-periphery/pull/8/commits/cafd3760ab444c56a6469a122f1a24eef74ad14a